x-pack/plugin/ml/src/test/java/org/elasticsearch/xpack/ml/queries/TextExpansionQueryBuilderTests.java

```diff
@@ -126,6 +139,46 @@ protected void doAssertLuceneQuery(TextExpansionQueryBuilder queryBuilder, Query
         }
     }
 
+    /**
+     * Overridden to ensure that {@link SearchExecutionContext} has a non-null {@link IndexReader}
+     */
+    @Override
+    public void testCacheability() throws IOException {
+        try (Directory directory = newDirectory(); RandomIndexWriter iw = new RandomIndexWriter(random(), directory)) {
+            Document document = new Document();
+            document.add(new FloatDocValuesField(RANK_FEATURES_FIELD, 1.0f));
+            iw.addDocument(document);
+            try (IndexReader reader = iw.getReader()) {
+                SearchExecutionContext context = createSearchExecutionContext(newSearcher(reader));
+                TextExpansionQueryBuilder queryBuilder = createTestQueryBuilder();
+                QueryBuilder rewriteQuery = rewriteQuery(queryBuilder, new SearchExecutionContext(context));
+
+                assertNotNull(rewriteQuery.toQuery(context));
+                assertTrue("query should be cacheable: " + queryBuilder.toString(), context.isCacheable());
+            }
+        }
+    }
+
+    /**
+     * Overridden to ensure that {@link SearchExecutionContext} has a non-null {@link IndexReader}
+     */
+    @Override
+    public void testToQuery() throws IOException {
```

Comment:
This is not right here, but I could use some guidance on how to correctly test this.

Right now it's failing because the TextExpansionQueryBuilder's doToQuery() method returns an IllegalStateException that this should have been written to another query type. I had to override this test, because if I don't this test fails because the SearchExecutionContext is null for this case. What's the best way to proceed for this type of situation?

CC: @benwtrent @jimczi

Reply:
OK, so I pushed some changes here, can you please let me know what you think? 
I pushed the lucene verification to the `WeightedTokensQueryBuilder` since `doToQuery` here always throws.

Reply:
@kderusso I think that is ok, but we need tests here to ensure we are rewriting to the correct thing.

Reply:
@benwtrent How do you recommend doing this if we're throwing on toQuery?

Reply:
having a test that calls rewrite with different objects and asserts on the result?

Reply:
Thanks, that's helpful. 

My issue here is that the `doRewrite()` method does an async inference call. I don't see any good examples of mocking this to get the rewrite that you're asking for. Would you be amenable to more in-depth yaml tests instead? If not, I think I'd benefit from understanding a bit more on what you're thinking here, thank you.